iaso/api/group_sets/serializers.py

```diff
@@ -0,0 +1,185 @@
+                debug_groups = [f"{g.name} ({g.id})" for g in groups]
+                raise serializers.ValidationError(
+                    detail={"group_ids": f"Some groups do not exist : found {debug_groups} vs {group_ids}."}
+                )
+
+            if self.instance and source_version_id is None:
+                source_version_id = self.instance.source_version.id
+
+            source_ids = list(set([group.source_version_id for group in groups]))
+            if len(source_ids) > 1:
+                raise serializers.ValidationError(
+                    detail={"group_ids": f"Groups do not all belong to the same SourceVersion : {source_ids}."}
+                )
+
+            if len(source_ids) == 1:
+                if source_ids[0] != source_version_id:
+                    raise serializers.ValidationError(
+                        detail={
+                            "group_ids": f"Groups do not all belong to the same as the groupset : {source_ids} vs {source_version_id}."
+                        }
+                    )
+
+        return attrs
+
+    def assign_relations(self, group_set):
+        request = self.context.get("request")
+        user = request.user
+
+        source_version_id = self.context["request"].data.get("source_version_id")
+        if source_version_id:
+            source_version = SourceVersion.objects.filter_for_user(user).get(pk=source_version_id)
+            if source_version:
+                group_set.source_version = source_version
+                group_set.save()
+
+        group_ids = self.context["request"].data.get("group_ids")
+        if group_ids:
+            # Ensure all provided group_ids are valid and belong to the same source_version
+            groups = Group.objects.filter(id__in=group_ids)
+
+            for g in groups:
+                group_set.groups.add(g)
+
+    def ensure_clean_validated_data(self, validated_data):
+        if "groups" in validated_data:
+            del validated_data["groups"]
+        if "group_ids" in validated_data:
+            del validated_data["group_ids"]
+
+    def create(self, validated_data):
+        self.ensure_clean_validated_data(validated_data)
+        group_set = GroupSet.objects.create(**validated_data)
+        self.assign_relations(group_set)
+        return group_set
+
+    def update(self, instance, validated_data):
+        original_copy = GroupSet.objects.get(pk=instance.id)
+
+        self.ensure_clean_validated_data(validated_data)
+        # patch behaviour
+        for attr, value in validated_data.items():
+            setattr(instance, attr, value)
+        instance.save()
+        self.assign_relations(instance)
+
+        request = self.context.get("request")
+        user = request.user
+
+        audit_models.log_modification(original_copy, instance, source=audit_models.GROUP_SET_API, user=user)
+
+        return instance
+
+    source_version = SourceVersionSerializerForGroupset(read_only=True)
```

Comment:
I would put the attributes above, before the `class Meta` so that it's easier to know what's in the serializer, but it's just personal preference